sdk/python/tests/test_job.py

```diff
@@ -0,0 +1,144 @@
+
+    @fixture
+    def avro_data_path(self):
+        final_results = tempfile.mktemp()
+        pandavro.to_avro(file_path_or_buffer=final_results, df=TEST_DATA_FRAME)
+        return final_results
+
+    def test_to_dataframe_local_file_staging_should_pass(
+        self, retrieve_job, avro_data_path, mocker
+    ):
+        mocker.patch.object(
+            retrieve_job.serving_stub,
+            "GetJob",
+            return_value=GetJobResponse(
+                job=BatchRetrievalJob(
+                    id="123",
+                    type=JobType.JOB_TYPE_DOWNLOAD,
+                    status=JobStatus.JOB_STATUS_DONE,
+                    file_uris=[f"file://{avro_data_path}"],
+                    data_format=DataFormat.DATA_FORMAT_AVRO,
+                )
+            ),
+        )
+        retrived_df = retrieve_job.to_dataframe()
+        assert_frame_equal(TEST_DATA_FRAME, retrived_df, check_like=True)
+
+    @mock_s3
+    def test_to_dataframe_s3_file_staging_should_pass(
+        self, retrieve_job, avro_data_path, mocker
+    ):
+        s3_client = boto3.client("s3")
+        target = "test_proj/test_features.avro"
+        s3_client.create_bucket(Bucket=BUCKET)
+        with open(avro_data_path, "rb") as data:
+            s3_client.upload_fileobj(data, BUCKET, target)
+
+        mocker.patch.object(
+            retrieve_job.serving_stub,
+            "GetJob",
+            return_value=GetJobResponse(
+                job=BatchRetrievalJob(
+                    id="123",
+                    type=JobType.JOB_TYPE_DOWNLOAD,
+                    status=JobStatus.JOB_STATUS_DONE,
+                    file_uris=[f"s3://{BUCKET}/{target}"],
+                    data_format=DataFormat.DATA_FORMAT_AVRO,
+                )
+            ),
+        )
+        retrived_df = retrieve_job.to_dataframe()
+        assert_frame_equal(TEST_DATA_FRAME, retrived_df, check_like=True)
+
+    @pytest.mark.parametrize(
+        "job_proto,exception",
+        [
+            (
+                GetJobResponse(
+                    job=BatchRetrievalJob(
+                        id="123",
+                        type=JobType.JOB_TYPE_DOWNLOAD,
+                        status=JobStatus.JOB_STATUS_DONE,
+                        data_format=DataFormat.DATA_FORMAT_AVRO,
+                        error="Testing job failure",
+                    )
+                ),
+                Exception,
+            ),
+            (
+                GetJobResponse(
+                    job=BatchRetrievalJob(
+                        id="123",
+                        type=JobType.JOB_TYPE_DOWNLOAD,
+                        status=JobStatus.JOB_STATUS_DONE,
+                        data_format=DataFormat.DATA_FORMAT_INVALID,
+                    )
+                ),
+                Exception,
+            ),
+        ],
+        ids=["when_retrieve_job_fails", "when_data_format_is_not_avro"],
+    )
+    def test_to_dataframe_s3_file_staging_should_raise(
```

Comment:
Might be out of scope for this PR but it would be nice if we can add test for GCS staging as well.

Reply:
@mrzzy for sure, I did try for gcs unit tests, I couldn't figure out the errors I was running into during gcs mocks, so I gave up.

Reply:
Lets track it over here. https://github.com/feast-dev/feast/issues/831